tools/x-repo-types/typeAnalyzer/typeAnalyzer.go

```diff
@@ -0,0 +1,575 @@
+// the `TypeNode.String()` method, and these values are ignored when building up the diff.
+// These exclusions represent types that are known to serialize in the same way, but
+// would be evaluated as different by the diff algorithm if not for this exclusion.
+var diffExclusions = map[string]bool{
+	// MicroAlgos is a struct with custom marshal override in go-algorand. In other repos it is a uint64.
+	`github.com/algorand/go-algorand/data/basics :: "basics.MicroAlgos" (struct)`: true,
+}
+
+// --------------- TYPE TREE DATA STRUCTURES --------------- //
+
+// TypeNode wraps reflect.Type and reflect.Kind to make it easier to
+// build a tree of types.
+type TypeNode struct {
+	Depth      int
+	Type       reflect.Type
+	Kind       reflect.Kind
+	ChildNames []ChildName
+	children   *ChildTypes
+}
+
+// String returns fully qualified information about the TypeNode.
+func (t *TypeNode) String() string {
+	return fmt.Sprintf("%s :: %q (%s)", t.Type.PkgPath(), t.Type, t.Kind)
+}
+
+// IsStruct returns true if the TypeNode is a struct.
+func (t *TypeNode) IsStruct() bool {
+	return t.Kind == reflect.Struct
+}
+
+// MakeType uses reflection to build a TypeNode from a concrete value.
+func MakeType(v interface{}) TypeNode {
+	t := reflect.TypeOf(v)
+	return TypeNode{Type: t, Kind: t.Kind()}
+}
+
+// ChildTypes represent a TypeNode's child Types.
+type ChildTypes map[string]TypeNode
+
+// ChildName represents the name a type is referred to together with
+// the Tag metadata (for the case of a struct field)
+type ChildName struct {
+	Name, Tag string
+}
+
+// String representing the ChildName in an invertible manner.
+func (e *ChildName) String() string {
+	return fmt.Sprintf("[%s](%s)", e.Name, e.Tag)
+}
+
+// ChildNameFromLabel de-"serialize"s a ChildName that has been rendered via ChildName.String()
+func ChildNameFromLabel(s string) (ChildName, error) {
+	re := regexp.MustCompile(`^\[(.*)\]\((.*)\)$`)
+	matches := re.FindStringSubmatch(s)
+	if len(matches) == 3 {
+		return ChildName{Name: matches[1], Tag: matches[2]}, nil
+	}
+	return ChildName{}, fmt.Errorf("invalid label: %s", s)
+}
+
+// Target represents a Type that is a child of another Type by
+// providing the ChildName that "points" to it.
+// In the case of a "root" TypeNode, it's up to the caller to provide
+// the ChildName.
+type Target struct {
+	ChildName
+	TypeNode TypeNode
+}
+
+// String is a convenience method for printing a Target.
+func (tgt *Target) String() string {
+	return fmt.Sprintf("%s|-->%s", &tgt.ChildName, &tgt.TypeNode)
+}
+
+// SerializationInfo provides the essential data
+// for go to serialize the field or other type.
+func (e *ChildName) SerializationInfo() string {
+	// Probably more subtlety is called for.
+	re := regexp.MustCompile(`^codec:"([^,"]+)`)
+
+	if e.Tag == "" {
+		return e.Name
+	}
+
+	matches := re.FindStringSubmatch(e.Tag)
+	if len(matches) > 1 {
+		return matches[1]
+	}
```

Comment:
I think you meant to use `ChildNameFromLabel`  here.

Reply:
Good catch, thx!